pkg/controllers/node/drift.go

```diff
@@ -0,0 +1,63 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package node
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/samber/lo"
+
+	"github.com/aws/karpenter-core/pkg/apis/config/settings"
+	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
+	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/utils/machine"
+)
+
+type Drift struct {
+	kubeClient    client.Client
+	cloudProvider cloudprovider.CloudProvider
+}
+
+func (d *Drift) Reconcile(ctx context.Context, provisioner *v1alpha5.Provisioner, node *v1.Node) (reconcile.Result, error) {
+	if !settings.FromContext(ctx).DriftEnabled {
+		return reconcile.Result{RequeueAfter: 5 * time.Minute}, nil
+	}
+	if _, ok := node.Labels[v1alpha5.ProvisionerNameLabelKey]; !ok {
+		return reconcile.Result{}, nil
+	}
+
+	if node.Annotations[v1alpha5.VoluntaryDisruptionAnnotationKey] == v1alpha5.VoluntaryDisruptionDriftedAnnotationValue {
```

Comment:
Don't we want to bail if this is set, regardless of the value? If some other disruption reason is set, we should respect it.

Reply:
Discussed offline, will remove check for value. There's still a case where we are patching instead of updating, where two concurrent patches will race and succeed, when only one should. Here's an issue to track that. https://github.com/aws/karpenter-core/issues/153